recipes/lora_finetune_distributed.py

```diff
@@ -277,45 +274,98 @@ def _setup_model(
-                model = config.instantiate(cfg_model)
+        with utils.set_default_dtype(self._dtype), torch.device("meta"):
+            model = config.instantiate(cfg_model)
 
-            log.info(
-                f"Model instantiation took {time.perf_counter() - init_start:.2f} secs"
+        if enable_activation_checkpointing:
+            utils.set_activation_checkpointing(
+                model, auto_wrap_policy={modules.TransformerDecoderLayer}
             )
 
+        for m in model.modules():
+            if isinstance(m, modules.TransformerDecoderLayer):
+                fully_shard(m)
+        fully_shard(model)
+        meta_sharded_sd = model.state_dict()
+        sharded_sd = {}
+
+        if self._is_rank_zero:
+
             # The model contains LoRA params which won't have any matching keys in
             # the state dict. As a result, we need to load with strict=False.
             # Before loading the state dict, ensure the state dict keys for the base
             # model and adapters (if available) match the keys in the full LoRA model
             # This is a good sanity check to prevent silent errors
-            validate_state_dict_for_lora(
-                lora_attn_modules=cfg_model.lora_attn_modules,
-                apply_lora_to_mlp=cfg_model.apply_lora_to_mlp,
-                apply_lora_to_output=getattr(cfg_model, "apply_lora_to_output", False),
-                full_model_state_dict_keys=model.state_dict().keys(),
-                lora_state_dict_keys=(
-                    lora_weights_state_dict.keys()
-                    if lora_weights_state_dict is not None
-                    else None
-                ),
-                base_model_state_dict_keys=base_model_state_dict.keys(),
-            )
+            # validate_state_dict_for_lora(
+            #     lora_attn_modules=cfg_model.lora_attn_modules,
+            #     apply_lora_to_mlp=cfg_model.apply_lora_to_mlp,
+            #     apply_lora_to_output=getattr(cfg_model, "apply_lora_to_output", False),
+            #     full_model_state_dict_keys=model.state_dict().keys(),
+            #     lora_state_dict_keys=(
+            #         lora_weights_state_dict.keys()
+            #         if lora_weights_state_dict is not None
+            #         else None
+            #     ),
+            #     base_model_state_dict_keys=base_model_state_dict.keys(),
+            # )
 
             # Load both the base model weights and (if available) the adapter weights. Both
             # of this should happen only on Rank 0
-            model.load_state_dict(base_model_state_dict, strict=False)
-            if lora_weights_state_dict:
-                model.load_state_dict(lora_weights_state_dict, strict=False)
-
+            log.info("FSDP is enabled. Loading checkpoints for Rank 0 ...")
+            init_start = time.perf_counter()
+            for param_name, full_param in base_model_state_dict.items():
```

Comment:
In the future, I think refactoring the conversion of CPU full state dict on rank 0 to sharded state dict on all ranks can be moved to a utility. (Having it here initially in the draft PR to show the difference sounds good though.)

Reply:
yep. will move this to utility since the same logic applies to lora state dict